contracts/utils/KyberInputScalingHelper.sol

```diff
@@ -0,0 +1,345 @@
+            return abi.encodeWithSelector(selector, callTarget, desc, targetData, clientData);
+        } else revert('InputScalingHelper: Invalid selector');
+    }
+
+    function _getScaledInputDataV2(
+        IMetaAggregationRouterV2.SwapDescriptionV2 memory desc,
+        bytes memory executorData,
+        uint256 newAmount,
+        bool isSimpleMode
+    ) internal pure returns (IMetaAggregationRouterV2.SwapDescriptionV2 memory, bytes memory) {
+        uint256 oldAmount = desc.amount;
+        if (oldAmount == newAmount) {
+            return (desc, executorData);
+        }
+
+        // simple mode swap
+        if (isSimpleMode) {
+            return (
+                _scaledSwapDescriptionV2(desc, oldAmount, newAmount),
+                _scaledSimpleSwapData(executorData, oldAmount, newAmount)
+            );
+        }
+
+        //normal mode swap
+        return (
+            _scaledSwapDescriptionV2(desc, oldAmount, newAmount),
+            _scaledExecutorCallBytesData(executorData, oldAmount, newAmount)
+        );
+    }
+
+    /// @dev Scale the swap description
+    function _scaledSwapDescriptionV2(
+        IMetaAggregationRouterV2.SwapDescriptionV2 memory desc,
+        uint256 oldAmount,
+        uint256 newAmount
+    ) internal pure returns (IMetaAggregationRouterV2.SwapDescriptionV2 memory) {
+        desc.minReturnAmount = (desc.minReturnAmount * newAmount) / oldAmount;
+        if (desc.minReturnAmount == 0) desc.minReturnAmount = 1;
+        desc.amount = newAmount;
+        for (uint256 i = 0; i < desc.srcReceivers.length; i++) {
+            desc.srcAmounts[i] = (desc.srcAmounts[i] * newAmount) / oldAmount;
+        }
+        return desc;
+    }
+
+    /// @dev Scale the executorData in case swapSimpleMode
+    function _scaledSimpleSwapData(
+        bytes memory data,
+        uint256 oldAmount,
+        uint256 newAmount
+    ) internal pure returns (bytes memory) {
+        SimpleSwapData memory swapData = abi.decode(data, (SimpleSwapData));
+        for (uint256 i = 0; i < swapData.firstPools.length; i++) {
+            swapData.firstSwapAmounts[i] = (swapData.firstSwapAmounts[i] * newAmount) / oldAmount;
+        }
+        return abi.encode(swapData);
+    }
+
+    /// @dev Scale the executorData in case normal swap
+    function _scaledExecutorCallBytesData(
+        bytes memory data,
+        uint256 oldAmount,
+        uint256 newAmount
+    ) internal pure returns (bytes memory) {
+        SwapExecutorDescription memory executorDesc = abi.decode(data, (SwapExecutorDescription));
+        executorDesc.minTotalAmountOut = (executorDesc.minTotalAmountOut * newAmount) / oldAmount;
+        for (uint256 i = 0; i < executorDesc.swapSequences.length; i++) {
+            Swap memory swap = executorDesc.swapSequences[i][0];
+            bytes4 functionSelector = swap.functionSelector;
+
+            if (functionSelector == IExecutorHelper.executeUniSwap.selector) {
```

Comment:
Do we have to update this contract when they add a new exchange?

Reply:
yes, but it wouldn't be breaking if we put in kyber API to only use these exchanges that are available in contract, and only add new after changing the contract